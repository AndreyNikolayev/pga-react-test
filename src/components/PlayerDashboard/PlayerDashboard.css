.PlayerDashboard .delete-btn {
    cursor: pointer;
    color: #d32f2f;
}

.PlayerDashboard .player-row:hover {
    cursor: pointer;
    background-color: #eee;
}