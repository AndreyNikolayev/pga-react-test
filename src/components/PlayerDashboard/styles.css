.player-dashboard .delete-btn {
    cursor: pointer;
    color: #d32f2f;
}

.player-dashboard .player-row:hover {
    cursor: pointer;
    background-color: #eee;
}