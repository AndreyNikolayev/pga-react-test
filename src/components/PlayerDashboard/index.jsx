import React, { Component } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { withRouter } from 'react-router';
import { deletePlayer } from '../../state/actions';
import { getPlayers } from '../../state/selectors';
import { Link} from 'react-router-dom';
import './PlayerDashboard.css';

class PlayerDashboard extends Component {

  formatName(player) {
    return `${player.lastName}, ${player.firstName}`
  }

  renderRows() {
    const { players } = this.props;

    if (!players) {
      return null
    }

    const playerRows = Object.keys(players).map((playerKey) => {
      return (
        <tr onClick={() => this.goToDetails(playerKey)} className="player-row" key={playerKey}>
          <td>{this.formatName(players[playerKey])}</td>
          <td>{players[playerKey].score}</td>
          <td><i onClick={(e) => this.remove(e, playerKey)} className="material-icons delete-btn">delete</i></td>
        </tr>
      )
    })
    return playerRows
  }

  goToDetails(playerKey) {
    this.props.history.push(`edit/${playerKey}`)
  }

  remove(event, playerKey) {
    event.stopPropagation()
    const { deletePlayer } = this.props

    deletePlayer(playerKey)
  }

  render() {
    return (
      <div className="PlayerDashboard">
        <h3 className="left">Leaderboard</h3>
        <Link to="/edit" className="waves-effect waves-light btn right" style={{ marginTop: 40 }}>Add Player</Link>
        <table className="table striped" style={{ marginTop: 20 }}>
          <thead>
            <tr>
              <th>Name</th>
              <th>Score</th>
              <th></th>
            </tr>
          </thead>
          <tbody>
            { this.renderRows() }
          </tbody>
        </table>
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  return {
    players: getPlayers(state)
  }
}
  
export default compose(
  withRouter,
  connect(mapStateToProps, {deletePlayer})
)(PlayerDashboard)