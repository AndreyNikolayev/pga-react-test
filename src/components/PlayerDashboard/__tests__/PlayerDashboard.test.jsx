import { shallow } from 'enzyme';
import React from 'react';
import PlayerDashboard from '..';

describe('PlayerDashboard', () => {

  let wrapper;
  let mockProps;
  const selectPlayerMock = jest.fn();
  const deletePlayerMock = jest.fn();

  beforeEach(() => {
    mockProps = {
      selectPlayer: selectPlayerMock,
      deletePlayer: deletePlayerMock,
      players: {
        '1' : {
          firstName: 'First 1',
          lastName: 'Last 1',
          score: 1
        },
        '2': {
          firstName: 'First 2',
          lastName: 'Last 2',
          score: 2
        }
      }
    };

    wrapper = shallow(<PlayerDashboard {...mockProps}/>);
  })

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has table with same amount of rows as players array', () => {
    expect(wrapper.find('tbody tr').length).toBe(Object.keys(mockProps.players).length);
  });

  it('handles row click', () => {
    const firstRow = wrapper.find('tbody tr').first();
    firstRow.simulate('click');

    expect(selectPlayerMock).toBeCalled();
    selectPlayerMock.mockReset();
  });

  it('handles delete click', () => {
    const firstDeleteButton = wrapper.find('tr .delete-btn').first();
    firstDeleteButton.simulate('click', {
      stopPropagation: () => {
      }
     });

    expect(deletePlayerMock).toBeCalled();
    expect(deletePlayerMock.mock.calls[0][0]).toBe(Object.keys(mockProps.players)[0]);
    deletePlayerMock.mockReset();
  });

  it('does not have rows if there are no players', () => {
    wrapper.setProps({ players: null });

    expect(wrapper.find('tbody tr').length).toBe(0);
  });
});