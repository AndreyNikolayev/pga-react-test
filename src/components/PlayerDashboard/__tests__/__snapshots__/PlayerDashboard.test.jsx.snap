// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerDashboard renders correctly 1`] = `
<div
  className="player-dashboard"
>
  <h3
    className="left"
  >
    Leaderboard
  </h3>
  <Link
    className="waves-effect waves-light btn right"
    replace={false}
    style={
      Object {
        "marginTop": 40,
      }
    }
    to="/edit"
  >
    Add Player
  </Link>
  <table
    className="table striped"
    style={
      Object {
        "marginTop": 20,
      }
    }
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Score
        </th>
        <th />
      </tr>
    </thead>
    <tbody>
      <tr
        className="player-row"
        key="1"
        onClick={[Function]}
      >
        <td>
          Last 1, First 1
        </td>
        <td>
          1
        </td>
        <td>
          <i
            className="material-icons delete-btn"
            onClick={[Function]}
          >
            delete
          </i>
        </td>
      </tr>
      <tr
        className="player-row"
        key="2"
        onClick={[Function]}
      >
        <td>
          Last 2, First 2
        </td>
        <td>
          2
        </td>
        <td>
          <i
            className="material-icons delete-btn"
            onClick={[Function]}
          >
            delete
          </i>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
